File: android/app/src/main/kotlin/app/coulombmppt/data/modbus/ModbusCrc.kt
package app.coulombmppt.data.modbus

// Standard Modbus RTU CRC-16. The MPPT firmware uses the exact same routine
// the JS app's cmdParse.js implements (poly 0xA001 = 40961, init 0xFFFF, LSB
// out first). Lifted verbatim into Kotlin so we can both sign outgoing
// frames and validate incoming ones.
object ModbusCrc {

    /** Returns the two CRC bytes in wire order — `[lo, hi]`. */
    fun compute(data: ByteArray, length: Int = data.size): ByteArray {
        var crc = 0xFFFF
        for (i in 0 until length) {
            crc = crc xor (data[i].toInt() and 0xFF)
            repeat(8) {
                crc = if ((crc and 0x0001) != 0) (crc ushr 1) xor 0xA001
                      else                       (crc ushr 1)
            }
        }
        val lo = (crc and 0x00FF).toByte()
        val hi = ((crc ushr 8) and 0x00FF).toByte()
        return byteArrayOf(lo, hi)
    }

    /**
     * Validate that the trailing two bytes of [frame] match a fresh CRC
     * computed over everything before them. Returns true on a valid frame.
     */
    fun verify(frame: ByteArray): Boolean {
        if (frame.size < 4) return false                       // slave + fn + crcLo + crcHi
        val crc = compute(frame, frame.size - 2)
        return crc[0] == frame[frame.size - 2] && crc[1] == frame[frame.size - 1]
    }
}

File: windows/src/CoulombMppt/Ble/ModbusCrc.cs
namespace CoulombMppt.Ble;

// Standard Modbus RTU CRC-16. The MPPT firmware uses the exact same routine
// the vendor JS app's cmdParse.js implements (poly 0xA001, init 0xFFFF, low
// byte out first). Ported verbatim from the Android client's ModbusCrc.kt so
// we can both sign outgoing frames and validate incoming ones.
public static class ModbusCrc
{
    /// <summary>Returns the two CRC bytes in wire order — [lo, hi].</summary>
    public static byte[] Compute(byte[] data, int length)
    {
        int crc = 0xFFFF;
        for (int i = 0; i < length; i++)
        {
            crc ^= data[i] & 0xFF;
            for (int b = 0; b < 8; b++)
            {
                crc = (crc & 0x0001) != 0
                    ? (crc >> 1) ^ 0xA001
                    : (crc >> 1);
            }
        }
        byte lo = (byte)(crc & 0x00FF);
        byte hi = (byte)((crc >> 8) & 0x00FF);
        return new[] { lo, hi };
    }

    public static byte[] Compute(byte[] data) => Compute(data, data.Length);

    /// <summary>
    /// Validate that the trailing two bytes of <paramref name="frame"/> match a
    /// fresh CRC computed over everything before them. True on a valid frame.
    /// </summary>
    public static bool Verify(byte[] frame)
    {
        if (frame.Length < 4) return false;          // slave + fn + crcLo + crcHi
        var crc = Compute(frame, frame.Length - 2);
        return crc[0] == frame[^2] && crc[1] == frame[^1];
    }
}
